kubetest/kops.go

```diff
@@ -483,3 +527,93 @@ func (k *kops) runKopsDump() (*kopsDump, error) {
 
 	return dump, nil
 }
+
+// getRandomAWSZones looks up all regions, and the availability zones for those regions.  A random
+// region is then chosen and the AZ's for that region is returned.  If HA is required a regions with 3
+// or more zones is returned.
+func getRandomAWSZones(isHA bool) (*[]string, error) {
+	ec2Session, err := getAWSEC2Session("us-east-2")
+
+	regionResults, err := ec2Session.DescribeRegions(&ec2.DescribeRegionsInput{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to call aws api DescribeRegions: %v", err)
+	}
+
+	var regions []string
+	regionMap := make(map[string][]string)
+
+	// we have to iterate over the regions to get the AZs for each region.
+	for _, r := range regionResults.Regions {
+		regions = append(regions, *r.RegionName)
+		ec2Session, err := getAWSEC2Session(*r.RegionName)
+
+		// az for each of the regions. AWS Go API does not allow us to make a single call
+		zonesResults, err := ec2Session.DescribeAvailabilityZones(&ec2.DescribeAvailabilityZonesInput{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to call aws api DescribeAvailabilityZones: %v", err)
+		}
+		// build a map of region and the regions az so that we can lookup AZ easily
+		for _, z := range zonesResults.AvailabilityZones {
+			existing, ok := regionMap[*z.RegionName]
+			var az []string
+			if ok {
+				az = append(existing, *z.ZoneName)
+			} else {
+				az = []string{*z.ZoneName}
+			}
+			regionMap[*z.RegionName] = az
+		}
+	}
+
+	// initialize global pseudo random generator
+	rand.Seed(time.Now().Unix())
+
+	// select a random region and zones
+	r := rand.Intn(len(regions))
+	selectedRegion := regions[r]
+	selectedZones := regionMap[selectedRegion]
+
+	// make sure the region has 3 or more AZ
+	if isHA {
+		for {
+
+			selectedZones = regionMap[selectedRegion]
+			if len(selectedZones) >= 3 {
+				break
+			}
+
+			// remove the region from the slice
+			regions[r] = regions[len(regions)-1]
+			regions = regions[:len(regions)-1]
+
+			if len(regions) == 0 {
+				return nil, fmt.Errorf("unable to find HA region")
+			}
+
+			rand.Seed(time.Now().Unix())
```

Comment:
I don't think you should be reseeding - or is there a trick here I'm missing?

Reply:
removed